src/shims/unix/fd.rs

```diff
@@ -619,22 +613,63 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let bytes = this.read_bytes_ptr_strip_provenance(buf, Size::from_bytes(count))?.to_owned();
         // We temporarily dup the FD to be able to retain mutable access to `this`.
         let Some(fd) = this.machine.fds.get(fd) else {
-            return Ok(Scalar::from_target_isize(this.fd_not_found()?, this));
-        };
-
-        let result = match offset {
-            None => fd.write(&fd, communicate, &bytes, this),
-            Some(offset) => {
-                let Ok(offset) = u64::try_from(offset) else {
-                    let einval = this.eval_libc("EINVAL");
-                    this.set_last_error(einval)?;
-                    return Ok(Scalar::from_target_isize(-1, this));
-                };
-                fd.pwrite(communicate, &bytes, offset, this)
-            }
+            let res = this.fd_not_found()?;
+            this.write_scalar(Scalar::from_target_isize(res, this), dest)?;
+            return Ok(());
         };
-
-        let result = result?.map(|c| i64::try_from(c).unwrap());
-        Ok(Scalar::from_target_isize(this.try_unwrap_io_result(result)?, this))
+        let res = write_byte_helper(offset, fd, bytes, communicate, this)?;
+        this.write_scalar(res, dest)?;
+        Ok(())
     }
 }
+fn read_byte_helper<'tcx>(
```

Comment:
Please add a doc comment explaining what this does.

Also, would be good to find a better name -- "helper" says basically nothing.